experimental/serde_unaligned/src/uvec.rs

```diff
@@ -0,0 +1,205 @@
+use crate::byte_slice::*;
+
+enum UVecInner<'a, T> {
+    Owned(Vec<T>),
+    Aligned(&'a [T]),
+    /// Buffer with little-endian data
+    UnalignedLE(&'a [u8]),
+    // Buffer with native-endian data (TODO)
+    // UnalignedNE(&'a [u8]),
+}
+
+/// A vector that may be borrowed and may point at unaligned data.
+pub struct UVec<'a, T>(UVecInner<'a, T>);
+
+impl<T> From<Vec<T>> for UVec<'_, T> {
+    fn from(vec: Vec<T>) -> Self {
+        Self(UVecInner::Owned(vec))
+    }
+}
+
+impl<'a, T> From<&'a [T]> for UVec<'a, T> {
+    fn from(slice: &'a [T]) -> Self {
+        Self(UVecInner::Aligned(slice))
+    }
+}
+
+impl<'a, T> UVec<'a, T> {
+    pub fn from_unaligned_le_bytes(bytes: &'a [u8]) -> Self {
+        Self(UVecInner::UnalignedLE(bytes))
+    }
+
+    pub fn from_unaligned_ne_bytes(_bytes: &'a [u8]) -> Self {
+        unimplemented!()
+    }
+}
+
+// TODO: I guess we can't implement std::ops::Index because that returns references, right?
+impl<'a, T> UVec<'a, T>
+where
+    T: Copy + From<ByteSliceLE<'a, 4>>,
```

Comment:
@Manishearth - how do I ask the compiler to use this implementation over all possible values of the `N` const generic?